src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/ValidationAttributeProvider.cs

```diff
@@ -0,0 +1,111 @@
+        private readonly IClientModelValidatorProvider _clientModelValidatorProvider;
+
+        /// <summary>
+        /// Initializes a new <see cref="ValidationAttributeProvider"/> instance.
+        /// </summary>
+        /// <param name="optionsAccessor">The accessor for <see cref="MvcViewOptions"/>.</param>
+        /// <param name="metadataProvider">The <see cref="IModelMetadataProvider"/>.</param>
+        /// <param name="clientValidatorCache">The <see cref="ClientValidatorCache"/> that provides
+        /// a list of <see cref="IClientModelValidator"/>s.</param>
+        public ValidationAttributeProvider(
+            IOptions<MvcViewOptions> optionsAccessor,
+            IModelMetadataProvider metadataProvider,
+            ClientValidatorCache clientValidatorCache)
+        {
+            if (optionsAccessor == null)
+            {
+                throw new ArgumentNullException(nameof(optionsAccessor));
+            }
+
+            if (metadataProvider == null)
+            {
+                throw new ArgumentNullException(nameof(metadataProvider));
+            }
+
+            if (clientValidatorCache == null)
+            {
+                throw new ArgumentNullException(nameof(clientValidatorCache));
+            }
+
+            _clientValidatorCache = clientValidatorCache;
+            _metadataProvider = metadataProvider;
+
+            var clientValidatorProviders = optionsAccessor.Value.ClientModelValidatorProviders;
+            _clientModelValidatorProvider = new CompositeClientModelValidatorProvider(clientValidatorProviders);
+        }
+
+        /// <inheritdoc />
+        public virtual void AddValidationAttributes(
+            ViewContext viewContext,
+            ModelExplorer modelExplorer,
+            string expression,
+            IDictionary<string, string> attributes)
+        {
+            if (viewContext == null)
+            {
+                throw new ArgumentNullException(nameof(viewContext));
+            }
+
+            if (attributes == null)
+            {
+                throw new ArgumentNullException(nameof(attributes));
+            }
+
+            // Only render attributes if client-side validation is enabled, and then only if we've
+            // never rendered validation for a field with this name in this form.
+            var formContext = viewContext.ClientValidationEnabled ? viewContext.FormContext : null;
+            if (formContext == null)
+            {
+                return;
+            }
+
+            var fullName = viewContext.ViewData.TemplateInfo.GetFullHtmlFieldName(expression);
+            if (formContext.RenderedField(fullName))
+            {
+                return;
+            }
+
+            formContext.RenderedField(fullName, true);
+
+            modelExplorer = modelExplorer ??
+                ExpressionMetadataProvider.FromStringExpression(expression, viewContext.ViewData, _metadataProvider);
```

Comment:
Discussed offline. I'll add a separate `AddAndTrackValidationAttributes()` method that wraps `AddValidationAttributes()`.

I'll leave the client-side validation check in `AddValidationAttributes()`. This has the minor downside that both methods will do this check but avoids changing the semantics of `FormContext.RenderedFields` i.e. the dictionary is never touched when client-side validation is disabled.

I'll also change `AddValidationAttributes()` to `throw` if `modelExplorer == null`. Not particularly clean to have such low-level code creating `ModelExplorer`s. I've been looking for opportunities to move such lines up a layer or two.